File: l33tcode/maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def widthOfBinaryTree(self, root: TreeNode) -> int:
        left = lambda x: x * 2 + 1
        right = lambda x: x * 2 + 2

        result = 0

        if not root:
            return result

        stack = [(root, 0)]

        while stack:
            result = max(stack[-1][1] - stack[0][1] + 1, result)

            old_stack = stack
            stack = []

            for node, pos in old_stack:
                if node.left:
                    stack.append((node.left, left(pos)))
                if node.right:
                    stack.append((node.right, right(pos)))

        return result

File: l33tcode/maximum-width-of-binary-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func widthOfBinaryTree(root *TreeNode) int {
	if root == nil {
		return 0
	}

	minOnLevel := map[int]int{}

	result := 1

	updateMin := func(level int, pos int) {
		_, ok := minOnLevel[level]

		if !ok {
			minOnLevel[level] = pos
		}

		if pos < minOnLevel[level] {
			minOnLevel[level] = pos
		}

		diff := pos - minOnLevel[level] + 1
		if diff > result {
			result = diff
		}
	}

	var dfs func(node *TreeNode, level int, pos int)
	dfs = func(node *TreeNode, level int, pos int) {
		if node.Left != nil {
			newPos := pos*2 + 1
			updateMin(level, newPos)
			dfs(node.Left, level+1, newPos)
		}
		if node.Right != nil {
			newPos := pos*2 + 2
			updateMin(level, newPos)
			dfs(node.Right, level+1, newPos)
		}
	}

	dfs(root, 0, 0)

	return result
}

func main() {
	root := &TreeNode{1, nil, nil}
	node := &TreeNode{2, nil, nil}
	root.Left = node
	widthOfBinaryTree(root)
}
